src/rgw/rgw_sal_motr.cc

```diff
@@ -2073,71 +2161,241 @@ int MotrObject::read_mobj(const DoutPrefixProvider* dpp, int64_t off, int64_t en
+                          M0_IC_GET, this->get_key().to_str(), bl);
+      ldpp_dout(dpp, 0) << __func__ << "*** have instance : cache miss : motr call : rc : " << rc << dendl;
+      if(rc < 0) {
+        ldpp_dout(dpp, 0) << __func__ << " ERROR: do_idx_op_by_name failed to get object's entry: rc="
+                          << rc << dendl;
+        return rc;
+      }
+      iter = bl.cbegin();
+      ent.decode(iter);
+      rc = 0;
+      // Put into the cache
+      this->store->get_obj_meta_cache()->put(dpp, this->get_name(), bl);
+      goto out;
     }
+    else
+    {
+      // Version-id instance is empty
+      // Cache miss.
+      keys[0] = this->get_name();
 
-    ldpp_dout(dpp, 20) <<__func__<< ": versioned bucket!" << dendl;
-    keys[0] = this->get_name();
-    rc = store->next_query_by_name(bucket_index_iname, keys, vals);
-    if (rc < 0) {
-      ldpp_dout(dpp, 0) << __func__ << "ERROR: NEXT query failed. " << rc << dendl;
-      return rc;
-    }
+      // Retrieve all 'max' number of pairs.
+      rc = store->next_query_by_name(bucket_index_iname, keys, vals, this->get_name());
+      ldpp_dout(dpp, 2) << "**** enabled : not instance : no null : rc : "<< rc << dendl;
+      if (rc < 0) {
+        ldpp_dout(dpp, 0) << __func__ << "ERROR: NEXT query failed. " << rc << dendl;
+        return rc;
+      }
 
-    rc = -ENOENT;
-    for (const auto& bl: vals) {
-      if (bl.length() == 0)
-        break;
+      //rc = -ENOENT;
 
-      iter = bl.cbegin();
-      ent_to_check.decode(iter);
-      if (ent_to_check.is_current()) {
-        ldpp_dout(dpp, 20) <<__func__<< ": found current version!" << dendl;
-        ent = ent_to_check;
-        rc = 0;
+      // Iterating on object keys and return the latest object version
+      int i = 1;
+      for (; i < rc; ++i) {
+      //for (const auto& bl: vals) {
+        if (vals[i].length() == 0)
+        {  ldpp_dout(dpp, 2) << "**** vals[i].length == 0 ? " << dendl; 
+          break;
+        }
+        iter = vals[i].cbegin();
+        ent.decode(iter);
+        ldpp_dout(dpp, 2) << "**** going to check for current version" << dendl;
+        if (ret_code == 0 && ent_to_check.key.instance == null_key.key.instance)
+        {
+          ldpp_dout(dpp, 2) << "**** inside if...no need to update is-latest flag" << dendl;
+          return ret_code;
+        }
 
-        this->store->get_obj_meta_cache()->put(dpp, this->get_name(), bl);
+        if (keys[i] == null_version_key)
+          continue;
 
-        break;
+        if (ent.is_current()) {
+          ldpp_dout(dpp, 20) <<__func__<< ": found current version!" << dendl;
+          rc = 0;
+          //ent.key.instance = "null";
+          // Put into the cache
+          this->store->get_obj_meta_cache()->put(dpp, this->get_name(), bl);
+          break;
+        }
+        this->store->get_obj_meta_cache()->put(dpp, this->get_key().to_str(), bl);
       }
-    }
-  } else {
+        // for (const auto& bl: vals) {
```

Comment:
why this commented out code is kept still?

Reply:
removed the commented code.
done